File: internal/validator/schema_validator.go
package validator

import (
	"encoding/json"
	"fmt"
	"regexp"
	"strings"
	"time"

	"github.com/karthik/anybase/pkg/models"
)

// SchemaValidator validates documents against OpenAPI schema
type SchemaValidator struct{}

// NewSchemaValidator creates a new schema validator
func NewSchemaValidator() *SchemaValidator {
	return &SchemaValidator{}
}

// ValidateDocument validates a document against a collection schema
func (v *SchemaValidator) ValidateDocument(doc map[string]interface{}, schema *models.CollectionSchema) error {
	if schema == nil {
		return nil // No schema, no validation
	}

	// Check required fields
	for _, required := range schema.Required {
		if _, ok := doc[required]; !ok {
			return fmt.Errorf("required field '%s' is missing", required)
		}
	}

	// Validate each property
	for key, value := range doc {
		// Skip internal fields
		if strings.HasPrefix(key, "_") {
			continue
		}

		prop, ok := schema.Properties[key]
		if !ok {
			// Check if additional properties are allowed
			if schema.AdditionalProperties == false {
				return fmt.Errorf("property '%s' is not allowed", key)
			}
			continue
		}

		if err := v.validateValue(value, prop, key); err != nil {
			return err
		}
	}

	return nil
}

// validateValue validates a value against a schema property
func (v *SchemaValidator) validateValue(value interface{}, prop *models.SchemaProperty, path string) error {
	if value == nil {
		// Check if null is allowed
		if prop.Type == "null" {
			return nil
		}
		if arr, ok := prop.Type.([]interface{}); ok {
			for _, t := range arr {
				if t == "null" {
					return nil
				}
			}
		}
		return fmt.Errorf("field '%s' cannot be null", path)
	}

	// Get the type(s)
	types := v.getTypes(prop.Type)
	
	// Try to validate against any of the allowed types
	var lastErr error
	for _, t := range types {
		if err := v.validateType(value, t, prop, path); err == nil {
			return nil // Valid for this type
		} else {
			lastErr = err
		}
	}
	
	if lastErr != nil {
		return lastErr
	}

	return fmt.Errorf("field '%s' does not match any allowed type", path)
}

// getTypes extracts types from type definition
func (v *SchemaValidator) getTypes(typeValue interface{}) []string {
	switch t := typeValue.(type) {
	case string:
		return []string{t}
	case []interface{}:
		types := []string{}
		for _, tt := range t {
			if str, ok := tt.(string); ok {
				types = append(types, str)
			}
		}
		return types
	default:
		return []string{}
	}
}

// validateType validates value against a specific type
func (v *SchemaValidator) validateType(value interface{}, dataType string, prop *models.SchemaProperty, path string) error {
	switch dataType {
	case "string":
		str, ok := value.(string)
		if !ok {
			return fmt.Errorf("field '%s' must be a string", path)
		}
		return v.validateString(str, prop, path)
		
	case "number", "integer":
		var num float64
		switch n := value.(type) {
		case float64:
			num = n
		case int:
			num = float64(n)
		case int64:
			num = float64(n)
		case int32:
			num = float64(n)
		default:
			return fmt.Errorf("field '%s' must be a number", path)
		}
		
		if dataType == "integer" {
			if num != float64(int64(num)) {
				return fmt.Errorf("field '%s' must be an integer", path)
			}
		}
		
		return v.validateNumber(num, prop, path)
		
	case "boolean":
		if _, ok := value.(bool); !ok {
			return fmt.Errorf("field '%s' must be a boolean", path)
		}
		
	case "array":
		arr, ok := value.([]interface{})
		if !ok {
			// Try to convert from other slice types
			if jsonBytes, err := json.Marshal(value); err == nil {
				if err := json.Unmarshal(jsonBytes, &arr); err != nil {
					return fmt.Errorf("field '%s' must be an array", path)
				}
			} else {
				return fmt.Errorf("field '%s' must be an array", path)
			}
		}
		return v.validateArray(arr, prop, path)
		
	case "object":
		obj, ok := value.(map[string]interface{})
		if !ok {
			return fmt.Errorf("field '%s' must be an object", path)
		}
		return v.validateObject(obj, prop, path)
	}
	
	return nil
}

// validateString validates string constraints
func (v *SchemaValidator) validateString(str string, prop *models.SchemaProperty, path string) error {
	// Check enum
	if len(prop.Enum) > 0 {
		found := false
		for _, e := range prop.Enum {
			if str == e {
				found = true
				break
			}
		}
		if !found {
			return fmt.Errorf("field '%s' must be one of: %v", path, prop.Enum)
		}
	}

	// Check pattern
	if prop.Pattern != "" {
		matched, err := regexp.MatchString(prop.Pattern, str)
		if err != nil {
			return fmt.Errorf("invalid pattern for field '%s': %v", path, err)
		}
		if !matched {
			return fmt.Errorf("field '%s' does not match pattern: %s", path, prop.Pattern)
		}
	}

	// Check length constraints
	if prop.MinLength != nil && len(str) < *prop.MinLength {
		return fmt.Errorf("field '%s' is too short (minimum length: %d)", path, *prop.MinLength)
	}
	if prop.MaxLength != nil && len(str) > *prop.MaxLength {
		return fmt.Errorf("field '%s' is too long (maximum length: %d)", path, *prop.MaxLength)
	}

	// Check format
	if prop.Format != "" {
		if err := v.validateFormat(str, prop.Format, path); err != nil {
			return err
		}
	}

	return nil
}

// validateNumber validates number constraints
func (v *SchemaValidator) validateNumber(num float64, prop *models.SchemaProperty, path string) error {
	// Check enum
	if len(prop.Enum) > 0 {
		found := false
		for _, e := range prop.Enum {
			if enumNum, ok := e.(float64); ok && num == enumNum {
				found = true
				break
			}
		}
		if !found {
			return fmt.Errorf("field '%s' must be one of: %v", path, prop.Enum)
		}
	}

	// Check range constraints
	if prop.Minimum != nil && num < *prop.Minimum {
		return fmt.Errorf("field '%s' is too small (minimum: %f)", path, *prop.Minimum)
	}
	if prop.Maximum != nil && num > *prop.Maximum {
		return fmt.Errorf("field '%s' is too large (maximum: %f)", path, *prop.Maximum)
	}

	return nil
}

// validateArray validates array constraints
func (v *SchemaValidator) validateArray(arr []interface{}, prop *models.SchemaProperty, path string) error {
	// Check length constraints
	if prop.MinItems != nil && len(arr) < *prop.MinItems {
		return fmt.Errorf("field '%s' has too few items (minimum: %d)", path, *prop.MinItems)
	}
	if prop.MaxItems != nil && len(arr) > *prop.MaxItems {
		return fmt.Errorf("field '%s' has too many items (maximum: %d)", path, *prop.MaxItems)
	}

	// Check unique items
	if prop.UniqueItems {
		seen := make(map[string]bool)
		for _, item := range arr {
			key := fmt.Sprintf("%v", item)
			if seen[key] {
				return fmt.Errorf("field '%s' must have unique items", path)
			}
			seen[key] = true
		}
	}

	// Validate each item
	if prop.Items != nil {
		for i, item := range arr {
			itemPath := fmt.Sprintf("%s[%d]", path, i)
			if err := v.validateValue(item, prop.Items, itemPath); err != nil {
				return err
			}
		}
	}

	return nil
}

// validateObject validates object constraints
func (v *SchemaValidator) validateObject(obj map[string]interface{}, prop *models.SchemaProperty, path string) error {
	// Check required fields
	for _, required := range prop.Required {
		if _, ok := obj[required]; !ok {
			return fmt.Errorf("required field '%s.%s' is missing", path, required)
		}
	}

	// Validate properties
	if prop.Properties != nil {
		for key, value := range obj {
			if subProp, ok := prop.Properties[key]; ok {
				subPath := fmt.Sprintf("%s.%s", path, key)
				if err := v.validateValue(value, subProp, subPath); err != nil {
					return err
				}
			}
		}
	}

	return nil
}

// validateFormat validates string format
func (v *SchemaValidator) validateFormat(str string, format string, path string) error {
	switch format {
	case "date-time":
		if _, err := time.Parse(time.RFC3339, str); err != nil {
			return fmt.Errorf("field '%s' must be a valid date-time (RFC3339 format)", path)
		}
	case "date":
		if _, err := time.Parse("2006-01-02", str); err != nil {
			return fmt.Errorf("field '%s' must be a valid date (YYYY-MM-DD format)", path)
		}
	case "email":
		emailRegex := `^[a-zA-Z0-9._%+-]+@[a-zA-Z0-9.-]+\.[a-zA-Z]{2,}$`
		if matched, _ := regexp.MatchString(emailRegex, str); !matched {
			return fmt.Errorf("field '%s' must be a valid email address", path)
		}
	case "uri", "url":
		urlRegex := `^(https?|ftp)://[^\s/$.?#].[^\s]*$`
		if matched, _ := regexp.MatchString(urlRegex, str); !matched {
			return fmt.Errorf("field '%s' must be a valid URL", path)
		}
	case "uuid":
		uuidRegex := `^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$`
		if matched, _ := regexp.MatchString(uuidRegex, strings.ToLower(str)); !matched {
			return fmt.Errorf("field '%s' must be a valid UUID", path)
		}
	}

	return nil
}

File: dashboard/lib/schema-validator.ts
export interface SchemaProperty {
  type?: string | string[];
  format?: string;
  description?: string;
  default?: any;
  enum?: any[];
  pattern?: string;
  minLength?: number;
  maxLength?: number;
  minimum?: number;
  maximum?: number;
  minItems?: number;
  maxItems?: number;
  uniqueItems?: boolean;
  items?: SchemaProperty;
  properties?: Record<string, SchemaProperty>;
  required?: string[];
  readOnly?: boolean;
  writeOnly?: boolean;
}

export interface Schema {
  type: string;
  properties?: Record<string, SchemaProperty>;
  required?: string[];
  additionalProperties?: boolean | SchemaProperty;
  description?: string;
}

export interface ValidationError {
  field: string;
  message: string;
}

export function validateDocument(document: any, schema?: Schema | null): ValidationError[] {
  const errors: ValidationError[] = [];
  
  if (!schema || !schema.properties) {
    return errors; // No schema, no validation
  }

  // Check required fields
  if (schema.required) {
    for (const field of schema.required) {
      if (!(field in document) || document[field] === undefined || document[field] === null || document[field] === '') {
        errors.push({
          field,
          message: `Field "${field}" is required`,
        });
      }
    }
  }

  // Check each property in the document
  for (const [key, value] of Object.entries(document)) {
    const property = schema.properties[key];
    
    if (!property) {
      // Check if additional properties are allowed
      if (schema.additionalProperties === false) {
        errors.push({
          field: key,
          message: `Field "${key}" is not allowed`,
        });
      }
      continue;
    }

    // Validate the property
    const propErrors = validateProperty(key, value, property);
    errors.push(...propErrors);
  }

  return errors;
}

function validateProperty(
  fieldPath: string,
  value: any,
  property: SchemaProperty
): ValidationError[] {
  const errors: ValidationError[] = [];

  // Get the expected type(s)
  const types = Array.isArray(property.type) 
    ? property.type 
    : property.type 
    ? [property.type] 
    : [];

  if (types.length === 0) return errors; // No type specified, skip validation

  // Allow null if it's in the type array
  if (value === null) {
    if (!types.includes('null')) {
      errors.push({
        field: fieldPath,
        message: `Field "${fieldPath}" cannot be null`,
      });
    }
    return errors;
  }

  // Check type
  const actualType = getJsonType(value);
  
  // Special case: 'number' type should accept both integers and floats
  let typeMatches = false;
  if (types.includes('number') && (actualType === 'number' || actualType === 'integer')) {
    typeMatches = true;
  } else if (types.includes(actualType)) {
    typeMatches = true;
  }
  
  if (!typeMatches) {
    errors.push({
      field: fieldPath,
      message: `Field "${fieldPath}" must be of type ${types.join(' or ')}, got ${actualType}`,
    });
    return errors; // Stop validation if type is wrong
  }

  // Type-specific validations
  switch (actualType) {
    case 'string':
      errors.push(...validateString(fieldPath, value, property));
      break;
    case 'number':
    case 'integer':
      errors.push(...validateNumber(fieldPath, value, property, actualType === 'integer'));
      break;
    case 'array':
      errors.push(...validateArray(fieldPath, value, property));
      break;
    case 'object':
      errors.push(...validateObject(fieldPath, value, property));
      break;
  }

  // Check enum values
  if (property.enum && property.enum.length > 0) {
    if (!property.enum.includes(value)) {
      errors.push({
        field: fieldPath,
        message: `Field "${fieldPath}" must be one of: ${property.enum.join(', ')}`,
      });
    }
  }

  return errors;
}

function validateString(
  fieldPath: string,
  value: string,
  property: SchemaProperty
): ValidationError[] {
  const errors: ValidationError[] = [];

  // Check length constraints
  if (property.minLength !== undefined && value.length < property.minLength) {
    errors.push({
      field: fieldPath,
      message: `Field "${fieldPath}" must be at least ${property.minLength} characters`,
    });
  }

  if (property.maxLength !== undefined && value.length > property.maxLength) {
    errors.push({
      field: fieldPath,
      message: `Field "${fieldPath}" must not exceed ${property.maxLength} characters`,
    });
  }

  // Check pattern
  if (property.pattern) {
    try {
      const regex = new RegExp(property.pattern);
      if (!regex.test(value)) {
        errors.push({
          field: fieldPath,
          message: `Field "${fieldPath}" does not match pattern: ${property.pattern}`,
        });
      }
    } catch (e) {
      // Invalid regex pattern
    }
  }

  // Check format
  if (property.format) {
    const formatErrors = validateFormat(fieldPath, value, property.format);
    errors.push(...formatErrors);
  }

  return errors;
}

function validateNumber(
  fieldPath: string,
  value: number,
  property: SchemaProperty,
  mustBeInteger: boolean
): ValidationError[] {
  const errors: ValidationError[] = [];

  if (mustBeInteger && !Number.isInteger(value)) {
    errors.push({
      field: fieldPath,
      message: `Field "${fieldPath}" must be an integer`,
    });
  }

  if (property.minimum !== undefined && value < property.minimum) {
    errors.push({
      field: fieldPath,
      message: `Field "${fieldPath}" must be at least ${property.minimum}`,
    });
  }

  if (property.maximum !== undefined && value > property.maximum) {
    errors.push({
      field: fieldPath,
      message: `Field "${fieldPath}" must not exceed ${property.maximum}`,
    });
  }

  return errors;
}

function validateArray(
  fieldPath: string,
  value: any[],
  property: SchemaProperty
): ValidationError[] {
  const errors: ValidationError[] = [];

  if (property.minItems !== undefined && value.length < property.minItems) {
    errors.push({
      field: fieldPath,
      message: `Field "${fieldPath}" must have at least ${property.minItems} items`,
    });
  }

  if (property.maxItems !== undefined && value.length > property.maxItems) {
    errors.push({
      field: fieldPath,
      message: `Field "${fieldPath}" must not exceed ${property.maxItems} items`,
    });
  }

  if (property.uniqueItems) {
    const uniqueValues = new Set(value.map(v => JSON.stringify(v)));
    if (uniqueValues.size !== value.length) {
      errors.push({
        field: fieldPath,
        message: `Field "${fieldPath}" must have unique items`,
      });
    }
  }

  // Validate each item
  if (property.items) {
    value.forEach((item, index) => {
      const itemErrors = validateProperty(
        `${fieldPath}[${index}]`,
        item,
        property.items!
      );
      errors.push(...itemErrors);
    });
  }

  return errors;
}

function validateObject(
  fieldPath: string,
  value: any,
  property: SchemaProperty
): ValidationError[] {
  const errors: ValidationError[] = [];

  if (property.properties) {
    // Check required fields
    if (property.required) {
      for (const field of property.required) {
        if (!(field in value)) {
          errors.push({
            field: `${fieldPath}.${field}`,
            message: `Field "${fieldPath}.${field}" is required`,
          });
        }
      }
    }

    // Validate each nested property
    for (const [key, val] of Object.entries(value)) {
      const nestedProp = property.properties[key];
      if (nestedProp) {
        const nestedErrors = validateProperty(
          `${fieldPath}.${key}`,
          val,
          nestedProp
        );
        errors.push(...nestedErrors);
      }
    }
  }

  return errors;
}

function validateFormat(
  fieldPath: string,
  value: string,
  format: string
): ValidationError[] {
  const errors: ValidationError[] = [];

  switch (format) {
    case 'email':
      const emailRegex = /^[^\s@]+@[^\s@]+\.[^\s@]+$/;
      if (!emailRegex.test(value)) {
        errors.push({
          field: fieldPath,
          message: `Field "${fieldPath}" must be a valid email address`,
        });
      }
      break;

    case 'date':
      // YYYY-MM-DD format
      const dateRegex = /^\d{4}-\d{2}-\d{2}$/;
      if (!dateRegex.test(value)) {
        errors.push({
          field: fieldPath,
          message: `Field "${fieldPath}" must be a valid date (YYYY-MM-DD)`,
        });
      }
      break;

    case 'date-time':
      // ISO 8601 format
      const dateTime = new Date(value);
      if (isNaN(dateTime.getTime())) {
        errors.push({
          field: fieldPath,
          message: `Field "${fieldPath}" must be a valid date-time`,
        });
      }
      break;

    case 'uri':
    case 'url':
      try {
        new URL(value);
      } catch {
        errors.push({
          field: fieldPath,
          message: `Field "${fieldPath}" must be a valid URL`,
        });
      }
      break;

    case 'uuid':
      const uuidRegex = /^[0-9a-f]{8}-[0-9a-f]{4}-[1-5][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$/i;
      if (!uuidRegex.test(value)) {
        errors.push({
          field: fieldPath,
          message: `Field "${fieldPath}" must be a valid UUID`,
        });
      }
      break;

    case 'ipv4':
      const ipv4Regex = /^(\d{1,3}\.){3}\d{1,3}$/;
      if (!ipv4Regex.test(value)) {
        errors.push({
          field: fieldPath,
          message: `Field "${fieldPath}" must be a valid IPv4 address`,
        });
      }
      break;

    case 'ipv6':
      const ipv6Regex = /^([\da-f]{1,4}:){7}[\da-f]{1,4}$/i;
      if (!ipv6Regex.test(value)) {
        errors.push({
          field: fieldPath,
          message: `Field "${fieldPath}" must be a valid IPv6 address`,
        });
      }
      break;
  }

  return errors;
}

function getJsonType(value: any): string {
  if (value === null) return 'null';
  if (Array.isArray(value)) return 'array';
  
  const type = typeof value;
  if (type === 'number') {
    return Number.isInteger(value) ? 'integer' : 'number';
  }
  
  return type;
}